### DISCUSSION TITLE
Call runtime API on top of any specified state

### DISCUSSION THREAD

**Discussion by liuchengxu** [paritytech/substrate#11095] [Ideas]
Currently, the runtime API is limited to be called on top of a certain block, while in Subspace we have to call some APIs against an intermediate state between the entire block execution, for example, [call `BlockBuilder_finalize_block` after all the extrinsics have been applied and before executing the `finalize_block` hook](https://github.com/subspace/subspace/blob/99ecf82b1b/cumulus/client/cirrus-executor/src/lib.rs#L649-L659). It's less robust having to work with the String API like `BlockBuilder_finalize_block` even it's kind of stable, I'm wondering whether there is a way to call the runtime API against any specific state, seemingly doable to do something around `contextual_call` https://github.com/paritytech/substrate/blob/6506784c95c2b6ee1db19cdbfc8142e9c7a071dc/client/service/src/client/call_executor.rs#L186

Not sure if you think it's a generally useful feature and could be included into Substrate, and any suggestions on implementing this feature outside Substrate would be appreciated!

**Comment by bkchr** [comment]
That is something that would be needed by Cumulus as well: https://github.com/paritytech/cumulus/issues/874

So, I think it can be integrated into Substrate :)

**Reply by liuchengxu** [reply]
Do you already have some plan in your mind that what it could be done properly in Substrate? Just in case I work on the wrong track :D.

**Reply by bkchr** [reply]
No I don't have yet a proper plan. Probably something around implementing `CallApiAt` for some new type.
